librz/include/rz_il/rzil_vm.h

```diff
@@ -0,0 +1,126 @@
+// SPDX-FileCopyrightText: 2021 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#ifndef RZIL_VM_H
+#define RZIL_VM_H
+
+#include "definitions/definitions.h"
+#include "rzil_opcodes.h"
+
+#define RZ_IL_VM_MAX_VAR  2048
+#define RZ_IL_VM_MAX_VAL  1024
+#define RZ_IL_VM_MAX_LAB  1024
+#define RZ_IL_VM_MAX_EFF  1024
+#define RZ_IL_VM_MAX_FLG  1024
+#define RZ_IL_VM_MAX_TEMP 32
+
+typedef struct rz_il_vm_t *RzILVM;
+typedef void (*RzILOpHandler)(RzILVM vm, RzILOp op);
+typedef void (*RzILVmHook)(RzILVM vm, RzILOp op);
+
+/**
+ *  \struct rz_il_vm_t
+ *  \brief core theory VM structure
+ */
+struct rz_il_vm_t {
+	RzILBag vm_global_value_set; ///< Store all RzILVal instance
+	RzILVar *vm_global_variable_list; ///< Store all RzILVar instance
+
+	RzILMem *mems; ///< Array of Memory, memory are actually hashmap in VM
+	int var_count, val_count, mem_count, lab_count; ///< count for VM predefined things
+	int addr_size; ///< size of address
+	int data_size; ///< size of minimal data unit
+
+	RzILTemp *temp_value_list; ///< Array of temporary values
+
+	HtPP *vm_global_bind_table; ///< Hashtable to record relationships between var and val
+	HtPP *vm_global_label_table; ///< Hashtable to maintain the label and address
+
+	HtPP *ct_opcodes; ///< Hashtable to maintain address and opcodes
+
+	RzILBitVector pc; ///< Program Counter of VM
+
+	RzILOpHandler *op_handler_table; ///< Array of Handler, handler can be indexed by opcode
+
+	int easy_debug; ///< Debug only, used to locate the bug
+};
+
+// VM operations about Variable and Value
+RZ_API RzILBitVector rz_il_hash_find_addr_by_lblname(RzILVM vm, const char *lbl_name);
```

Comment:
all these APIs are not unit tested... hard to see if they work well in theory and in practice.